### PR TITLE
PyTorch Image Models Integration Plan

### LINKED ISSUE #7045: Design timm integration
Write a design document detailing the plan to integrate timm in Geti. All components - UI, backend and library - shall be included in the design scope.

### PR DESCRIPTION
**Pull request by kprokofi**
### Summary

Resolves #7045

### Checklist

- [ ] The PR title and description are clear and descriptive
- [ ] I have manually tested the changes
- [ ] All changes are covered by automated tests
- [ ] All related issues are linked to this PR (if applicable)
- [ ] Documentation has been updated (if applicable)

### CODE REVIEW

File: library/docs/design/timm-integration.md
Comment (leoll2) [inline library/docs/design/timm-integration.md:1]: I'd suggest to move this file to the existing `library/docs/design/` folder.

File: library/docs/design/timm-integration.md [hunk 1]
```diff
@@ -0,0 +1,656 @@
+# Proposal: Integrate the Full `timm` Classification Catalog via Adaptive Model Manifests
+
+- Issue: [#5230 — Integrate pytorch-image-models (timm) for classification](https://github.com/open-edge-platform/geti/issues/5230)
+- Milestone: Geti 3.1
```
Comment (leoll2) [inline library/docs/design/timm-integration.md:4]:
```suggestion
```
The milestone is already attached to the issue, I'd remove this line to prevent inconsistencies if we later need to adjust the release scope.

File: docs/proposals/timm-integration.md [hunk 2]
```diff
@@ -0,0 +1,656 @@
...
+
+### 5.2 UI configuration behavior
+
+The selector stores the concrete architecture id. Changing the selection invalidates the previous training-configuration query and requests the newly selected id, so the backend builds a new dynamic manifest and returns new defaults.
+
+### 5.3 Testing
+
+- Library unit/model tests: for representative models from CNN and transformer families, verify the `timm.optim.create_optimizer_v2` options, then instantiate → 1-step train → export (OpenVINO/ONNX).
+- Backend unit tests: `model_name ↔ id` round-trip; `TimmManifestProvider.build_manifest` produces a schema-valid `ModelManifest` (respecting `extra="forbid"`); lazy `get_model_manifest_by_id` for timm ids; `GetiConfigConverter._resolve_recipe` + LR-injection fallback; `/timm/backbones` search/pagination/facets against a fixture snapshot.
+- E2E testing: each run should execute a small rotating sample of random timm architectures plus one basic baseline architecture, so we continuously validate representative coverage without trying to test all 1400 models every run.
```
Comment (leoll2) [inline docs/proposals/timm-integration.md]: This approach looks reasonable to me. I would add one requirement: when sampling the random architectures to test, we should guarantee that every family is represented with at least 1 model.

File: library/docs/design/timm-integration.md [hunk 3]
```diff
@@ -0,0 +1,656 @@
...
+
+- A generic timm backbone already exists: `TimmBackbone` (`src/getitune/backend/lightning/models/classification/backbones/timm.py`) wraps `timm.create_model(...)` and accepts any `model_name: str`.
+- Task models already exist and accept `model_name`:
+  - `TimmModelMulticlassCls` and `TimmModelMultilabelCls` (`.../classification/{multiclass,multilabel}_models/timm_model.py`).
+  - A `TimmModel` factory (`.../classification/factory.py`) dispatches by task and exposes `TimmModel.list_models()` → `timm.list_models(pretrained=True)`
+
+Blockers at scale:
+
+- One recipe YAML per model does not scale to 1400.
+- `TimmBackbone` currently hardcodes `num_classes=1000`, sets `self.model.classifier = None`, and relies on `forward_features` + `GlobalAveragePooling(dim=2)`. This assumes a CNN-style 2D feature map and a `.classifier`-named head. Many timm families (ViT/DeiT/BEiT/Swin/hybrids) use different head names and non-spatial outputs, so they will fail or silently mispreprocess → poor/zero accuracy. This is the single biggest correctness risk.
```
Comment (leoll2) [inline library/docs/design/timm-integration.md:114]: > ``TimmBackbone` currently hardcodes `num_classes=1000`

FYI this is already going to change in #6982, setting `num_classes=0`

File: library/docs/design/timm-integration.md [hunk 4]
```diff
@@ -0,0 +1,656 @@
...
+
+```yaml
+# library/src/getitune/recipe/classification/multi_class_cls/timm_generic.yaml
+task: MULTI_CLASS_CLS
+model:
+  class_path: getitune.backend.lightning.models.classification.multiclass_models.timm_model.TimmModelMulticlassCls
+  init_args:
+    label_info: 1000
+    model_name: tf_efficientnetv2_s.in21k   # <-- overwritten per user selection
+    learning_rate: 0.001                     # <-- overwritten from manifest/user (see §3.1)
```
Comment (leoll2) [inline library/docs/design/timm-integration.md]:
```suggestion
    model_name: DYNAMIC   # <-- overwritten per user selection
    learning_rate: DYNAMIC                     # <-- overwritten from manifest/user (see §3.1)
```
I suggest to use a fixed placeholder value for fields that need to be overwritten, e.g. "DYNAMIC". This way, if e.g. the `model_name` field is not correctly interpolated before the training starts, it will raise an error instead of falling back to another `tf_efficientnetv2_s.in21k`.

File: library/docs/design/timm-integration.md [hunk 5]
```diff
@@ -0,0 +1,656 @@
...
+- Rendering: a flat CSS `Grid` of radio-selectable cards inside one `RadioGroup` (`model-architectures-list-layout.component.tsx`).
+- Hyperparameters ("Training configuration") are configured in `train-model/advanced-settings/` and fetched per selected architecture via `GET /api/projects/{project_id}/training_configuration?model_architecture_id=…`. The panel renders generically: it walks the returned `parameters` tree and chooses a widget from each param's `value_type`/`allowed_values` (`ParameterField`). Nothing is hardcoded by name. Implication: whatever fields the backend serves render automatically — no new UI widgets are needed for LR/epochs/etc.
+
+Key insight: the `id → recipe` / `id → manifest` relationships are 1:1 static files today. Making timm a single parametrized recipe + a dynamically generated, per-id manifest removes the scaling problem. The UI treats the concrete architecture id (`image-classification-timm-<arch>`) as the selection key, so the existing per-id config/training pipeline is reused unchanged.
+
+---
+
+## 3. Required Library Changes
+
+### 3.1 Generic timm recipes
```
Comment (leoll2) [inline library/docs/design/timm-integration.md:144]: What about the existing models that are based on timm? Would you recommend keeping them or deprecating in favor of the generic recipe? If we decide to drop them, of course, we shouldn't forget about backward-compatibility.

File: library/docs/design/timm-integration.md [hunk 6]
```diff
@@ -0,0 +1,656 @@
...
+  "backbones": [
+    {
+      "model_name": "tf_efficientnetv2_s.in21k",
+      "family": "efficientnet",
+      "variant": "v2_s.in21k",
+      "input_size": [3, 300, 300],   // from pretrained_cfg
+      "mean": [0.485, 0.456, 0.406],
+      "std": [0.229, 0.224, 0.225],
+      "interpolation": "bicubic",
+      "default_lr": 0.007,           // aligned with the optimizer the library will pick (SGD)
```
Comment (leoll2) [inline library/docs/design/timm-integration.md:258]: What about all the other hyperparameters usually found in a manifest, e.g. `training.scheduler`, `dataset_preparation.augmentation`, ... Are their default values also specified through the catalog snapshot?

File: library/docs/design/timm-integration.md [hunk 7]
```diff
@@ -0,0 +1,656 @@
...
+    ModelManifestDeprecationStatus,
+)
+from app.models.training_configuration import AlgoLevelParameters, AlgoLevelTrainingParameters
+from app.models import TaskType
+
+_ID_PREFIX = "image-classification-timm-"
+
+def model_name_to_id(model_name: str) -> str:
+    # timm names contain '.', '_' — encode to a URL/DB-safe id, reversibly.
+    return _ID_PREFIX + model_name.replace(".", "--").replace("_", "-")
```
Comment (leoll2) [inline library/docs/design/timm-integration.md]: This sanitization step looks unnecessary: both `.` and `_` are perfectly safe characters for URLs and SQLite.

File: library/docs/design/timm-integration.md [hunk 8]
```diff
@@ -0,0 +1,656 @@
...
+            id=model_name_to_id(model_name),
+            name=f"timm · {model_name}",
+            family=e["family"],                 # new schema field (§5)
+            variant=e.get("variant"),           # new schema field (§5)
+            task=TaskType.CLASSIFICATION,
+            description=f"timm backbone '{model_name}'.",
+            pretrained_weights=None,             # weights are timm-managed (§6)
+            weights_source="timm_managed",      # new schema field (§6)
+            support_status=ModelManifestDeprecationStatus.ACTIVE,
+            capabilities=Capabilities(xai=True, tiling=False),
```
Comment (leoll2) [inline library/docs/design/timm-integration.md:368]: Is there any possibility to support tiling with timm models? And is it true that they support XAI? From my understanding, `timm.create_model` doesn't instantiate a XAI head.

File: docs/proposals/timm-integration.md [hunk 9]
```diff
@@ -0,0 +1,656 @@
...
+Two endpoints:
+
+4.7.1 Main list - one synthetic timm entry (the card). `GET /api/model_architectures?task=classification` returns the curated models plus a single synthetic entry so the UI can render the timm card:
+
+```jsonc
+{
+  "id": "image-classification-timm",       // synthetic entry point (not trainable directly)
+  "name": "Custom backbone (timm)",
+  "family": "timm",
+  "is_backbone_catalog": true,             // UI flag → render searchable selector
```
Comment (leoll2) [inline docs/proposals/timm-integration.md]: This field is redundant, you already return `"family": "timm"`: the UI can read this value and render the card accordingly.

File: library/docs/design/timm-integration.md [hunk 10]
```diff
@@ -0,0 +1,656 @@
...
+  "task": "classification",
+  "description": "Choose any of 1400+ timm backbones.",
+  "capabilities": { "xai": true, "tiling": false }
+}
+```
+
+4.7.2 New backbone-search endpoint - feeds the selector. Backed entirely by the snapshot - cheap, paginated, filterable:
+
+```
+GET /api/model_architectures/timm/backbones?search=&family=&page=&page_size=
```
Comment (leoll2) [inline library/docs/design/timm-integration.md]:
```suggestion
GET /api/model_architectures/timm/backbones?family=&limit=&offset=
```
For consistency with the other endpoints that use pagination

File: library/docs/design/timm-integration.md [hunk 11]
```diff
@@ -0,0 +1,656 @@
...
+```jsonc
+{
+  "id": "image-classification-timm",       // synthetic entry point (not trainable directly)
+  "name": "Custom backbone (timm)",
+  "family": "timm",
+  "is_backbone_catalog": true,             // UI flag → render searchable selector
+  "support_status": "active",
+  "task": "classification",
+  "description": "Choose any of 1400+ timm backbones.",
+  "capabilities": { "xai": true, "tiling": false }
```
Comment (leoll2) [inline library/docs/design/timm-integration.md:525]: I think the response should also include all the family names, for the UI to display a list of available choices.

**Review by leoll2 — COMMENTED** [review]
Solid document, I agree with most of the design choices. I would maybe add a short section about optimization and deployment: are all timm-based models exportable to OpenVINO out-of-the-box (hence compatible with ModelAPI-based inference) and optimizable through NNCF?

File: library/docs/design/timm-integration.md (hunk 8 above)
Comment (kprokofi) [reply library/docs/design/timm-integration.md:368]: We are in charge of the head, we can include XAI. Tiling is not needed for classification task and doesn't depend on the model TBH. It is False since it often doesn't need for classification

File: library/docs/design/timm-integration.md (hunk 6 above)
Comment (kprokofi) [reply library/docs/design/timm-integration.md:258]: augmentations are all supported for classification as in base.yaml, no need to store it here. For scheduler, yes, probably we need to add it in snapshot... Or we can try to use ReduceOnPlateue globally.

File: library/docs/design/timm-integration.md (hunk 5 above)
Comment (kprokofi) [reply library/docs/design/timm-integration.md:144]: We won't drop it. However, I think we should refactor and clean implementation of some of them. Especially VIT. But it is internal library refactoring, we will not remove separate models, moreover I opt to add new SOTA models from timm repo as separate templates as well with curated set of hyper parameters

File: docs/proposals/timm-integration.md (hunk 2 above)
Comment (kprokofi) [reply docs/proposals/timm-integration.md]: Yes, we can create a set for testing to include 1 model per family (maybe the smallest to save time)

File: library/docs/design/timm-integration.md (hunk 1 above)
Comment (kprokofi) [reply library/docs/design/timm-integration.md:4]: Sure

File: library/docs/design/timm-integration.md
Comment (kprokofi) [reply library/docs/design/timm-integration.md:1]: changes are not only for library. But I can move it there

File: library/docs/design/timm-integration.md (hunk 10 above)
Comment (kprokofi) [reply library/docs/design/timm-integration.md]: Will do

File: docs/proposals/timm-integration.md (hunk 9 above)
Comment (kprokofi) [reply docs/proposals/timm-integration.md]: Agree

**Review by leoll2 — APPROVED** [review]
